crates/nu-protocol/src/eval_const.rs

```diff
@@ -259,39 +263,51 @@ pub(crate) fn create_nu_constant(engine_state: &EngineState, span: Span) -> Valu
     Value::record(record, span)
 }
 
-pub fn get_vendor_autoload_dir(engine_state: &EngineState) -> Option<PathBuf> {
+pub fn get_vendor_autoload_dir(_engine_state: &EngineState) -> Vec<PathBuf> {
+    // TODO: update docs
     // pseudo code
     // if env var NU_VENDOR_AUTOLOAD_DIR is set, in any platform, use it
     // if not, if windows, use ALLUSERPROFILE\nushell\vendor\autoload
     // if not, if non-windows, if env var PREFIX is set, use PREFIX/share/nushell/vendor/autoload
     // if not, use the default /usr/share/nushell/vendor/autoload
 
-    // check to see if NU_VENDOR_AUTOLOAD_DIR env var is set, if not, use the default
-    Some(
-        option_env!("NU_VENDOR_AUTOLOAD_DIR")
-            .map(String::from)
-            .unwrap_or_else(|| {
-                if cfg!(windows) {
-                    let all_user_profile = match engine_state.get_env_var("ALLUSERPROFILE") {
-                        Some(v) => format!(
-                            "{}\\nushell\\vendor\\autoload",
-                            v.coerce_string().unwrap_or("C:\\ProgramData".into())
-                        ),
-                        None => "C:\\ProgramData\\nushell\\vendor\\autoload".into(),
-                    };
-                    all_user_profile
-                } else {
-                    // In non-Windows environments, if NU_VENDOR_AUTOLOAD_DIR is not set
-                    // check to see if PREFIX env var is set, and use it as PREFIX/nushell/vendor/autoload
-                    // otherwise default to /usr/share/nushell/vendor/autoload
-                    option_env!("PREFIX").map(String::from).map_or_else(
-                        || "/usr/local/share/nushell/vendor/autoload".into(),
-                        |prefix| format!("{}/share/nushell/vendor/autoload", prefix),
-                    )
-                }
-            })
-            .into(),
-    )
+    let into_autoload_path_fn = |mut path: PathBuf| {
+        path.push("nushell");
+        path.push("vendor");
+        path.push("autoload");
+        path
+    };
+    let mut dirs = Vec::new();
+
+    // platform base directies
+    // TODO: We're handling windows and not windows at the moment. using XDG_DATA_DIRS and falling back to `/usr/(local/)share` on MacOS might not be correct from my reading of the Apple developer documentation.
+    #[cfg(not(target_os = "windows"))]
+    // TODO: env var from engine state? How does that handle strings where a path is not valid unicode for some reason? Even if we do that here, the env vars will not be read from the engine state but the process in the `dirs_next` call below.
+    std::env::var_os("XDG_DATA_DIRS")
```

Comment:
I doubt this will be allowed. It's preferred to get env var from stack or engine_state.

Reply:
Well, that's why I asked that question in a todo comment right above. There's certain benefits to using `std::env::var_os`, such as also working with paths that are not valid unicode, and anything using `dirs_next` will use not us the stack or engine state anyway. I *can* switch to the engine_state if that's preferrable, but I'm just not sure on the trade-off here, and the inconsistency it'll introduce.

Reply:
As I recall, we try not to call into the OS environment at all in order to try and keep our thread-safety intact. Maybe other core-team members have opinions here.

Reply:
We just generally don't support non-Unicode paths. It's a known issue with how we do things, but it's probably not going to be resolved in this PR.

I think it is okay to use the true environment variable at load time, though, as we never set these and it is not possible to change the vendor autoload dir from Nushell configuration (that just wouldn't make sense).

We do need to load env variables from `Stack` at runtime but this is a bit of a different situation

Reply:
Yeah, `setup_config`, `get_vendor_autoload_dir`, and `read_vendor_autoload_files` should all only be called at shell startup (before any user code is run), so it's fine to use the process's environment variables instead of using the engine state and stack.